pynbody/snapshot/gadgethdf.py

```diff
@@ -154,6 +158,302 @@ class _SubfindHdfMultiFileManager(_GadgetHdfMultiFileManager):
+            # keep track of the slice for each gadget particle type.
+
+            ptype_slice_start = family_slice_start
+
+            for particle_type in self._family_to_group_map[fam]:
+
+                ptype_slice_len = 0
+                self._file_interrupt_points[particle_type] = []
+                for hdf_group in self._hdf_files.iter_particle_groups_with_name(particle_type):
+                    ptype_slice_len += hdf_group[self._hdf_files._size_from_hdf5_key].size
+                    self._file_interrupt_points[particle_type].append(ptype_slice_len)
+                self._file_ptype_slice[particle_type] = slice(ptype_slice_start, ptype_slice_start + ptype_slice_len)
+                family_length += ptype_slice_len
+                ptype_slice_start += ptype_slice_len
+
+            self._file_family_slice[fam] = slice(family_slice_start, family_slice_start + family_length)
+            family_slice_start += family_length
+
+        self._num_file_particle = family_slice_start
+        
+    def __init_load_map(self, take = None):
+        """ Set up family slice and particle count for loading """
+
+        self._load_control = chunk.LoadControl(self._file_ptype_slice, _max_buf, take) # use HDF groups type instead of family type here
+        self._family_slice_to_load = {}
+        self._num_particles_to_load = self._load_control.mem_num_particles
+
+        family_slice_start = 0
+        for fam in self._all_families:
+            family_length = 0
+
+            ptype_slice_start = family_slice_start
+
+            for particle_type in self._family_to_group_map[fam]:
+
+                ptype_slice_len = self._load_control.mem_family_slice[particle_type].stop - self._load_control.mem_family_slice[particle_type].start
+
+                family_length += ptype_slice_len
+                ptype_slice_start += ptype_slice_len
+
+            self._family_slice_to_load[fam] = slice(family_slice_start, family_slice_start + family_length)
+            family_slice_start += family_length
+        
+
+    def load_arrays(self, all_fams_to_load: list[family.Family], sim: SimSnap, array_name: str, translated_names: list[str]):
+        """Load an array from the HDF files into the simulation snapshot.
+        
+        Parameters
+        ----------
+        all_fams_to_load : list of pynbody.family.Family
+            A list of pynbody families for which to load the array.
+        sim : SimSnap
+            The parent SimSnap object.
+        array_name : str
+            The pynbody standard name for the array to load.
+        translated_names : list of str
+            A list of possible names for the array in the Gadget HDF file.
+
+        """
+
+        array_filler = _HDFArrayFiller()
+        for loading_fam in all_fams_to_load:
+            
+            sim_fam_array = sim[loading_fam][array_name]
+            array_filler._update_sim_element_size(sim_fam_array)
```

Comment:
it would seem to make sense to instantiate `_HDFArrayFiller` here, with `sim_fam_array` as an argument to the constructor -- again eliminating the possibility of using it in an uninitialized state